### PR TITLE
BUG, SIMD: Fix memory overlap in ufunc comparison loops

### LINKED ISSUE #22841: np.logical_xor.accumulate fails on 1.24 on mac
### Describe the issue:

.Accumulate on logical_xor gives wrong result on newest version. I suspect the same for bitwise_xor.

### Reproduce the code example:

```python
import numpy as np

a = np.array([False, False,  True, False, False, False, False,  True, False, False,  True, False, False,  True, False, False, False, False, False, False])
np.logical_xor.accumulate(a)
array([False, False,  True,  True, False, False, False,  True,  True,
       False,  True,  True, False,  True,  True, False, False, False,
       False, False])
# Should be [False, False,  True,  True, True, True, True, False,,,,]
```

### Error message:

_No response_

### NumPy/Python version information:

numpy=1.24
python 3.8, 
mac version 12.5.1

### Context for the issue:

_No response_

### PR DESCRIPTION
**Pull request by seiko2plus**
closes #22841, relates #21483

### CODE REVIEW

**Comment by seberg** [comment]
Isn't the problem that the `IS_BLOCKABLE_BINARY_BOOL` check is missing the `(abs_ptrdiff(args[2], args[1]) >= (vsize)` and similar checks that `IS_BLOCKABLE_BINARY_SCALAR1` and `IS_BLOCKABLE_BINARY_SCALAR2` have?

I guess this is fine as well, the more precise check probably doesn't matter in practice, but should we remove the macros if we don't expand the `IS_BLOCKABLE_BINARY_BOOL` to include the `vsize` based check?

**Comment by seberg** [comment]
This is confusing with the additional `_BOOL` versions!  How about removing the `_BOOL` macros and using the normal macros instead?  Maybe I am missing something, but it seems to me that should work?

**Comment by seberg** [comment]
Arg, sorry that was off, since the bool version deals with the output type being different.  But, I do think something like this would also work:
```diff
diff --git a/numpy/core/src/umath/fast_loop_macros.h b/numpy/core/src/umath/fast_loop_macros.h
index 5274957f7..973e47850 100644
--- a/numpy/core/src/umath/fast_loop_macros.h
+++ b/numpy/core/src/umath/fast_loop_macros.h
@@ -381,15 +381,25 @@ abs_ptrdiff(char *a, char *b)
 #define IS_BLOCKABLE_BINARY_BOOL(esize, vsize) \
     (steps[0] == (esize) && steps[0] == steps[1] && steps[2] == (1) && \
      npy_is_aligned(args[1], (esize)) && \
-     npy_is_aligned(args[0], (esize)))
+     npy_is_aligned(args[0], (esize)) && \
+     (abs_ptrdiff(args[2], args[0]) >= (vsize) || \
+      abs_ptrdiff(args[2], args[0]) == 0) && \
+     (abs_ptrdiff(args[2], args[1]) >= (vsize) || \
+      abs_ptrdiff(args[2], args[1]) >= 0))
 
 #define IS_BLOCKABLE_BINARY_SCALAR1_BOOL(esize, vsize) \
     (steps[0] == 0 && steps[1] == (esize) && steps[2] == (1) && \
-     npy_is_aligned(args[1], (esize)))
+     npy_is_aligned(args[1], (esize)) && \
+     ((abs_ptrdiff(args[2], args[1]) >= (vsize)) || \
+      (abs_ptrdiff(args[2], args[1]) == 0)) && \
+     abs_ptrdiff(args[2], args[0]) >= (esize))
 
 #define IS_BLOCKABLE_BINARY_SCALAR2_BOOL(esize, vsize) \
     (steps[0] == (esize) && steps[1] == 0 && steps[2] == (1) && \
-     npy_is_aligned(args[0], (esize)))
+     npy_is_aligned(args[0], (esize)) && \
+     ((abs_ptrdiff(args[2], args[0]) >= (vsize)) || \
+      (abs_ptrdiff(args[2], args[0]) == 0)) && \
+     abs_ptrdiff(args[2], args[1]) >= (esize))
 
 /* align var to alignment */
 #define LOOP_BLOCK_ALIGN_VAR(var, type, alignment)\
```

In any case either seems fine to me if you have a preference for this version.  But lets delete the faulty macro's if we go this route?

**Comment by seiko2plus** [comment]
I don't like to hide memory overlap tests behind macros, just one check at the beginning of the function and then using contiguous macros or going raw sounds safe and clean to me, we already follow this way for most of our dispatchable sources, therefore I deleted these macros since its no longer used anywhere.

**Comment by seberg** [comment]
Yeah, I can see a point in not hiding them in a macro that is used for other things especially.  Lets put this in then, it is indeed a pattern used elsewhere.

Thanks for the quick fix and sleuthing in what was wrong!

File: numpy/core/tests/test_umath.py [hunk 1]
```diff
@@ -31,6 +31,13 @@
     uf for uf in UFUNCS_UNARY if 'f->f' in uf.types
 ]
 
+UFUNCS_BINARY = [
+    uf for uf in UFUNCS if uf.nin == 2
+]
+UFUNCS_BINARY_ACC = [
+    uf for uf in UFUNCS_BINARY if hasattr(uf, "accumulate") and uf.nout == 1
```
Comment (seberg) [inline numpy/core/tests/test_umath.py:38]: Btw. the `hasattr` shouldn't do anything right now.  Checking here that `ufunc.signature is None` would make more sense to me.
But its mainly a nitpick and the rest is important for backporting.

Comment (seiko2plus) [reply numpy/core/tests/test_umath.py:38]: Thanks for clarification, in that case, the following skip can be avoided:

https://github.com/numpy/numpy/blob/1a0ea79eae7da0dad90051bd3c5d7a8097b427f2/numpy/core/tests/test_umath.py#L4356-L4357

**Pull request merged**